metrics/resource_view_test.go

```diff
@@ -189,6 +196,69 @@ func TestAllMetersExpiration(t *testing.T) {
 	// (123=9m, 456=evicted, 789=0m)
 }
 
+func TestIfAllMeterResourcesAreRemoved(t *testing.T) {
+	allMeters.clock = clock.Clock(clock.NewFakeClock(time.Now()))
+	var fakeClock *clock.FakeClock = allMeters.clock.(*clock.FakeClock)
+	ClearMetersForTest() // t+0m
+	// Register many resources at once
+	for i := 1; i <= 100; i++ {
+		res := resource.Resource{Labels: map[string]string{"foo": "bar"}}
+		res.Labels["id"] = fmt.Sprint(i)
+		if _, err := optionForResource(&res); err != nil {
+			t.Error("Should succeed getting option, instead got error ", err)
+		}
+		m := stats.Int64("testView_sum", "", stats.UnitDimensionless)
+		v := view.View{Name: fmt.Sprintf("testview-%d", i), Measure: m, Aggregation: view.Sum()}
+
+		err := RegisterResourceView(&v)
+		if err != nil {
+			t.Fatal("RegisterResourceView =", err)
+		}
+	}
+	allMeters.lock.Lock()
+	// Make a copy to test against as allMeters should be cleaned up
+	copyAllMeters := make(map[string]*meterExporter)
+	for k, v := range allMeters.meters {
+		copyAllMeters[k] = v
+	}
+	allMeters.lock.Unlock()
+	resourceViews.lock.Lock()
```

Comment:
[the comment on `resourceViews`](https://github.com/knative/pkg/blob/ba2137c16f4f4eb562b523e72f7267a6f6d4e3d6/metrics/resource_view.go#L61) says:

> // Lock regime: lock allMeters before resourceViews. The critical path is in
 // optionForResource, which must lock allMeters, but only needs to lock
 // resourceViews if a new meter needs to be created.

.. which makes me wonder if this locking is strictly correct. Maybe the allMeters lock should be held until after we unlock resourceViews.lock?

Reply:
Will change to be on the safe side. I have it this way here because I dont expect the views to be modified by some other entity.